### PR TITLE
containerize: pin the Spack version used in a container

### PR DESCRIPTION
**Pull request by alalazo**
This PR permits to specify the `url` and `ref` of the Spack instance used in a container recipe simply by expanding the YAML schema as outlined in #20442:
```yaml
container:
  images:
    os: amazonlinux:2
    spack:
      ref: develop
      resolve_sha: true
```
The `resolve_sha` option, if true, verifies the `ref` by cloning the Spack repository in a temporary directory and transforming any tag or branch name to a commit sha. When this new ability is leveraged an additional "bootstrap" stage is added, which builds an image with Spack setup and ready to install software. The Spack repository to be used can be customized with the `url` keyword under `spack`.

Modifications:
- [x] Permit to pin the version of Spack, either by branch or tag or sha
- [x] Added a few new OSes (centos:8, amazonlinux:2, ubuntu:20.04, alpine:3, cuda:11.2.1)
- [x] Permit to print the bootstrap image as a standalone
- [x] Add documentation on the new part of the schema
- [x] Add unit tests for different use cases

### CODE REVIEW

**Comment by haampie** [comment]
Some issues I found:

```
spack:
  view: true
  specs:
  - sirius +cuda ^cuda@:10.2 ^mpich ^openblas@0.3.10 threads=openmp target=haswell
  container:
    images:
      os: amazonlinux:2
      spack:
        ref: develop
        enforce-sha: true
```

- I'm used to clingo and I'm missing it here :p normally i wouldnt have to specify the openblas version and let the concretizer figure that one out
- `strip` fails because it is applied to powerpc shared libs that apparently ship with the cuda package
  ```
  Step 16/23 : RUN find -L /opt/view/* -type f -exec readlink -f '{}' \; |     xargs file -i |     grep 'charset=binary' |     grep 'x-executable\|x-archive\|x-sharedlib' |     awk -F: '{print $1}' | xargs strip -s
   ---> Running in 257649bd6d9a
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libcuda-injection.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/TreeLauncherSubreaper'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libInterceptorInjectionTarget.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libTreeLauncherTargetInjection.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libnvperf_target.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libnvperfapi64.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/nv-nsight-cu-cli'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libnvperf_host.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libNvSwCounterApi.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libTreeLauncherPlaceholder.so'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/TreeLauncherTargetLdPreloadHelper'
  strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-haswell/gcc-7.3.1/cuda-10.2.89-mao6gttdiesgc5ff4uvjdlk5iqodjubz/nsight-compute-2019.5.0/target/linux-desktop-glibc_2_19_0-ppc64le/libTreeLauncherTargetUpdatePreloadInjection.so'
  ```
- setting `LD_LIBRARY_PATH` makes system executables fail to work:
  ```
  bash-4.2# yum install libgomp
  There was a problem importing one of the Python modules
  required to run yum. The error leading to this problem was:
     /opt/view/lib/liblzma.so.5: version `XZ_5.1.2alpha' not found (required by /lib64/librpmio.so.3)
  Please install a package which provides this module, or
  verify that the module is installed correctly.
  It's possible that the above module doesn't match the
  current version of Python, which is:
  2.7.18 (default, Feb 18 2021, 06:07:59) 
  [GCC 7.3.1 20180712 (Red Hat 7.3.1-12)]
  If you cannot solve this problem yourself, please go to 
  the yum faq at:
    http://yum.baseurl.org/wiki/Faq
  ```

**Comment by alalazo** [comment]
Thanks for test-driving the PR @haampie

**Comment by alalazo** [comment]
@victorusu

File: lib/spack/docs/containers.rst
Comment (tgamblin) [inline lib/spack/docs/containers.rst]: This is the simplest available method of selecting base images

File: lib/spack/docs/containers.rst
Comment (tgamblin) [inline lib/spack/docs/containers.rst]: `url` specifies the URL from which to clone Spack. It defaults to https://github.com/spack/spack

File: lib/spack/docs/containers.rst
Comment (tgamblin) [inline lib/spack/docs/containers.rst]: Can we call this `resolve_shas` or something like that?  it's not really "enforcing" the sha -- it's translating other refs to commit shas.

File: lib/spack/docs/containers.rst
Comment (tgamblin) [inline lib/spack/docs/containers.rst]: probably needs a little clarifying -- "resolve branches and tags in `spack.yaml` to commits in the generated recipe"

File: lib/spack/llnl/util/filesystem.py
Comment (tgamblin) [inline lib/spack/llnl/util/filesystem.py]: moves -> cd's

maybe we should use this in more places...

File: lib/spack/spack/cmd/containerize.py
Comment (tgamblin) [inline lib/spack/spack/cmd/containerize.py]: phase -> stage?

File: lib/spack/spack/cmd/containerize.py
Comment (tgamblin) [inline lib/spack/spack/cmd/containerize.py]: phase -> stage?

File: lib/spack/spack/cmd/containerize.py
Comment (tgamblin) [inline lib/spack/spack/cmd/containerize.py]: I think this should be one argument like:

`--stages=LIST`

and default to `--stages=bootstrap,build,run`.  That, or make it `--last-stage=STAGE` and have an argument for the endpoint.  I think the first reads better, but the last prevents users from doing crazy things, though honestly you kind of have to know what you're doing to use this argument, so I think the `--stages` is fine.

File: share/spack/templates/container/amazonlinux_2.dockerfile
Comment (tgamblin) [inline share/spack/templates/container/amazonlinux_2.dockerfile]: where do we actually call `which` in spack scripts?  I think we should start preferring `command -v`, as that's actually a builtin.  Can we change that instead of including this in images?

File: lib/spack/spack/container/images.json
Comment (tgamblin) [inline lib/spack/spack/container/images.json:7]: This is a larger topic, but is it worth storing the list of OS packages here instead of in the templates?  One kind of interesting idea is that we could have a command for Spack to install them itself, using the OS package manager, at least until Spack knows how to bootstrap itself using only builtin python capabilities (which I'd like to eventually enable)

This would also decouple things from the OS and reduce the number of templates needed -- you could allow a user to specify any base image, even if it doesn't match the ones here, so long as they provided the necessary dependencies themselves.  But, if Spack detects an OS it knows how to bootstrap, it could install what it needs.

File: lib/spack/spack/schema/container.py
Comment (tgamblin) [inline lib/spack/spack/schema/container.py]: _ or -?

File: share/spack/templates/container/Dockerfile
Comment (tgamblin) [inline share/spack/templates/container/Dockerfile]: can this be a property or something instead of a hard-coded thing?  e.g `if render_phase.build` is easier to produce from the command line arg I mentioned above.

File: share/spack/templates/container/centos_7.dockerfile
Comment (tgamblin) [inline share/spack/templates/container/centos_7.dockerfile]: is this or Lmod directly needed?  Are people using modules w/spack containerize?  Could these ones just be optional?  I guess this is only for the bootstrap image but still.

**Review by tgamblin — CHANGES_REQUESTED** [review]
A few minor requests and some arg tweaks, and one larger question about bootstrapping images.

File: share/spack/templates/container/centos_7.dockerfile
Comment (alalazo) [reply share/spack/templates/container/centos_7.dockerfile]: Here I replicated what we currently do for our images on Dockerhub. The idea is that in a following PR we can remove the Dockerfile under `share/spack/docker` and generate them with containerize. In any case I agree that modules are not strictly needed for Spack to work correctly, so I'll try to trim the list of packages here.

File: lib/spack/spack/schema/container.py
Comment (alalazo) [reply lib/spack/spack/schema/container.py]: I think `_` it's better for consistency with other configuration option. Good catch.

File: lib/spack/spack/cmd/containerize.py
Comment (alalazo) [reply lib/spack/spack/cmd/containerize.py]: I'll go with `--last-stage` as it's safer and maps better to the existing implementation

File: share/spack/templates/container/amazonlinux_2.dockerfile
Comment (alalazo) [reply share/spack/templates/container/amazonlinux_2.dockerfile]: I'll check what can be trimmed, but for reference I started from:

https://github.com/spack/spack/blob/6b12000a7c81ca26b874b087a1a3c88034cfdd34/share/spack/docker/centos-7.dockerfile#L16-L34

which is what we currently build on Dockerhub

Comment (alalazo) [reply share/spack/templates/container/amazonlinux_2.dockerfile]: I'll check what can be trimmed, but for reference I started from:

https://github.com/spack/spack/blob/6b12000a7c81ca26b874b087a1a3c88034cfdd34/share/spack/docker/centos-7.dockerfile#L16-L34

which is what we currently build on Dockerhub

File: share/spack/templates/container/centos_7.dockerfile
Comment (alalazo) [reply share/spack/templates/container/centos_7.dockerfile]: See ab4533c

File: share/spack/templates/container/amazonlinux_2.dockerfile
Comment (alalazo) [reply share/spack/templates/container/amazonlinux_2.dockerfile]: See ab4533c

File: lib/spack/spack/container/images.json
Comment (alalazo) [reply lib/spack/spack/container/images.json:7]: I like this idea. Would it be fine with you to work on it in a separate PR?

**Comment by alalazo** [comment]
@tgamblin This is ready for another review

**Comment by alalazo** [comment]
@tgamblin ping

**Comment by alalazo** [comment]
@tgamblin Do you have some time for a second review?

**Comment by kniec** [comment]
@alalazo how can I help to push this forward?

**Comment by alalazo** [comment]
@kniec I'm just waiting for a review for this PR. In the meanwhile I'm trying to keep it up to date with `develop`.

**Comment by kniec** [comment]
@alalazo How to I specify `os_packages_final` in the spack.yaml?

**Comment by alalazo** [comment]
@kniec It should still be as in the example shown [here](https://spack.readthedocs.io/en/latest/containers.html#creating-images-from-environments):
```yaml
os_packages:
  final:
  - pkg1
  - pkg2
```

**Comment by kniec** [comment]
Thx, for AL2, the `epel-release` install leads to an error. 

```
No package epel-release available.
Error: Nothing to do
The command '/bin/sh -c yum update -y && yum install -y epel-release && yum update -y  && yum install -y libgomp  && rm -rf /var/cache/yum  && yum clean all' returned a non-zero code: 1
```

Can we remove `&& yum install -y epel-release && yum update -y` or replace it with `amazon-linux-extras install epel -y`

**Comment by kniec** [comment]
When doing the strip with CUDA.
```
Step 16/26 : RUN find -L /opt/view/* -type f -exec readlink -f '{}' \; |     xargs file -i |     grep 'charset=binary' |     grep 'x-executable\|x-archive\|x-sharedlib' |     awk -F: '{print $1}' | xargs strip -s
 ---> Running in 1ce5d8dd9d02
strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-skylake_avx512/gcc-7.3.1/cuda-11.4.0-jsrmv2nrjqqactjo77hglkzuy3iv22sq/nsight-compute-2021.2.0/target/linux-desktop-glibc_2_19_0-ppc64le/TreeLauncherSubreaper'
strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-skylake_avx512/gcc-7.3.1/cuda-11.4.0-jsrmv2nrjqqactjo77hglkzuy3iv22sq/nsight-compute-2021.2.0/target/linux-desktop-glibc_2_19_0-ppc64le/TreeLauncherTargetLdPreloadHelper'
strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-skylake_avx512/gcc-7.3.1/cuda-11.4.0-jsrmv2nrjqqactjo77hglkzuy3iv22sq/nsight-compute-2021.2.0/target/linux-desktop-glibc_2_19_0-ppc64le/libInterceptorInjectionTarget.so'
strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-skylake_avx512/gcc-7.3.1/cuda-11.4.0-jsrmv2nrjqqactjo77hglkzuy3iv22sq/nsight-compute-2021.2.0/target/linux-desktop-glibc_2_19_0-ppc64le/libTreeLauncherPlaceholder.so'
strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-skylake_avx512/gcc-7.3.1/cuda-11.4.0-jsrmv2nrjqqactjo77hglkzuy3iv22sq/nsight-compute-2021.2.0/target/linux-desktop-glibc_2_19_0-ppc64le/libTreeLauncherTargetInjection.so'
strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-skylake_avx512/gcc-7.3.1/cuda-11.4.0-jsrmv2nrjqqactjo77hglkzuy3iv22sq/nsight-compute-2021.2.0/target/linux-desktop-glibc_2_19_0-ppc64le/libTreeLauncherTargetUpdatePreloadInjection.so'
strip: Unable to recognise the format of the input file `/opt/software/linux-amzn2-skylake_avx512/gcc-7.3.1/cuda-11.4.0-jsrmv2nrjqqactjo77hglkzuy3iv22sq/nsight-compute-2021.2.0/target/linux-desktop-glibc_2_19_0-ppc64le/libcuda-injection.so'
```
Ah, already referenced before: https://github.com/spack/spack/pull/21910#issuecomment-789788688

**Comment by alalazo** [comment]
I fixed the epel-release issue on amazonlinux, and rebased. So this is again ready for a review :grimacing:

**Comment by kniec** [comment]
I updated the [containers-on-pcluster](https://containers-on-pcluster.workshop.aws/gromacs/build/single-node.html#c5) workshop to reflect that. Do you also are going to kick out the CUDA binary problem from above?

**Comment by alalazo** [comment]
@kniec Maybe in a smaller, follow up PR? Would that work?

**Comment by kniec** [comment]
@alalazo sure thing - I just kick that part out with
```bash
spack containerize \
  | sed -e '/RUN find -L /,+4d' > Dockerfile
```

**Comment by alalazo** [comment]
@vsoch With respect to Ubuntu 20.04 you can try:
```yaml
spack:
  specs:
  - zlib

  container:
    images:
      os: ubuntu:20.04
      spack:
        ref: develop
```
with this PR

**Comment by vsoch** [comment]
will do!

**Comment by vsoch** [comment]
Okay testing for 20.04! The Dockerfile looks great, builds without issue:

```
FROM ubuntu:20.04 as bootstrap

ENV SPACK_ROOT=/opt/spack \
    CURRENTLY_BUILDING_DOCKER_IMAGE=1 \
    container=docker

ENV DEBIAN_FRONTEND=noninteractive   \
    LANGUAGE=en_US.UTF-8 \
    LANG=en_US.UTF-8 \
    LC_ALL=en_US.UTF-8

RUN apt-get -yqq update \
 && apt-get -yqq install --no-install-recommends \
        build-essential \
        ca-certificates \
        curl \
        file \
        g++ \
        gcc \
        gfortran \
        git \
        gnupg2 \
        iproute2 \
        locales \
        lua-posix \
        make \
        python3 \
        python3-pip \
        python3-setuptools \
        unzip \
 && locale-gen en_US.UTF-8 \
 && pip3 install boto3 \
 && rm -rf /var/lib/apt/lists/*

RUN mkdir $SPACK_ROOT && cd $SPACK_ROOT && \
    git clone https://github.com/spack/spack.git . && git checkout develop  && \
    mkdir -p $SPACK_ROOT/opt/spack

RUN ln -s $SPACK_ROOT/share/spack/docker/entrypoint.bash \
          /usr/local/bin/docker-shell \
 && ln -s $SPACK_ROOT/share/spack/docker/entrypoint.bash \
          /usr/local/bin/interactive-shell \
 && ln -s $SPACK_ROOT/share/spack/docker/entrypoint.bash \
          /usr/local/bin/spack-env

RUN mkdir -p /root/.spack \
 && cp $SPACK_ROOT/share/spack/docker/modules.yaml \
        /root/.spack/modules.yaml \
 && rm -rf /root/*.* /run/nologin $SPACK_ROOT/.git

# [WORKAROUND]
# https://superuser.com/questions/1241548/
#     xubuntu-16-04-ttyname-failed-inappropriate-ioctl-for-device#1253889
RUN [ -f ~/.profile ]                                               \
 && sed -i 's/mesg n/( tty -s \\&\\& mesg n || true )/g' ~/.profile \
 || true


WORKDIR /root
SHELL ["docker-shell"]

# Creates the package cache
RUN spack spec hdf5+mpi

ENTRYPOINT ["/bin/bash", "/opt/spack/share/spack/docker/entrypoint.bash"]
CMD ["interactive-shell"]

# Build stage with Spack pre-installed and ready to be used
FROM bootstrap as builder


# What we want to install and how we want to install it
# is specified in a manifest file (spack.yaml)
RUN mkdir /opt/spack-environment \
&&  (echo "spack:" \
&&   echo "  specs:" \
&&   echo "  - zlib" \
&&   echo "  concretization: together" \
&&   echo "  config:" \
&&   echo "    install_tree: /opt/software" \
&&   echo "  view: /opt/view") > /opt/spack-environment/spack.yaml

# Install the software, remove unnecessary deps
RUN cd /opt/spack-environment && \
    spack env activate . && \
    spack install --fail-fast && \
    spack gc -y

# Strip all the binaries
RUN find -L /opt/view/* -type f -exec readlink -f '{}' \; | \
    xargs file -i | \
    grep 'charset=binary' | \
    grep 'x-executable\|x-archive\|x-sharedlib' | \
    awk -F: '{print $1}' | xargs strip -s

# Modifications to the environment that are necessary to run
RUN cd /opt/spack-environment && \
    spack env activate --sh -d . >> /etc/profile.d/z10_spack_environment.sh

# Bare OS image to run the installed executables
FROM ubuntu:20.04

COPY --from=builder /opt/spack-environment /opt/spack-environment
COPY --from=builder /opt/software /opt/software
COPY --from=builder /opt/view /opt/view
COPY --from=builder /etc/profile.d/z10_spack_environment.sh /etc/profile.d/z10_spack_environment.sh

ENTRYPOINT ["/bin/bash", "--rcfile", "/etc/profile", "-l"]
```

Shelling in is very straight forward:

```
$ docker run -it test 
```

I think my one question is if spack should be already on the path? As a user I wouldn't know what to do after shelling in and not seeing it.
```
root@a6cc2bd49210:/# which spack
```

**Comment by alalazo** [comment]
@vsoch The Dockerfile, by default, is a 3 stage build where Spack is in the first two. The last stage is just the software required in the environment. You can:
```console
$ spack containerize --last-stage build
```
If you want to stop before the last stage.

**Comment by vsoch** [comment]
ah okay, that does make sense. Sorry didn't look closely enough!

**Comment by vsoch** [comment]
@davidbeckingsale see this PR here for ubuntu 20.04 support

**Comment by kniec** [comment]
@alalazo `docker build` defines the stage to build for with `--target` - I am voting to use the same option.

**Comment by alalazo** [comment]
@kniec I think "target" would be confusing for people using Spack since in specs it's used to identify the target microarchitecture.

**Comment by kniec** [comment]
yeah, that is true.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
